### PR TITLE
[Android] Implement mute/unmute events for remote video tracks

### PR DESCRIPTION
**Pull request by saghul**
When a remote video track is not rendering any video it will emit the
muted event. When video recovers it will go back to the unmuted state.

This is what Chrome currently does for remote video tracks.

Ref:
https://chromium.googlesource.com/chromium/src.git/+/master/content/renderer/media/video_track_adapter.cc

### CODE REVIEW

**Review by paweldomas — APPROVED** [review]

**Comment by saghul** [comment]
@paweldomas @lyubomir PTAL. I changed some things since last time you reviewed:

- Added iOS implementation
- Raised timeouts (default ones seemed a bit too low)
- Simplified timer scheduling on Android

**Comment by lyubomir** [comment]
@paweldomas, could you please review and merge? (Because Saúl is going to try to update us to the latest WebRTC this week.)

**Comment by paweldomas** [comment]
> (Because Saúl is going to try to update us to the latest WebRTC this week.)

Should I do this before or after the update ?

**Comment by lyubomir** [comment]
> Should I do this before or after the update ?

@paweldomas, before the update please. So that we don't get into conflicts.

File: ios/RCTWebRTC/WebRTCModule+RTCPeerConnection.m [hunk 1]
```diff
@@ -377,22 +396,18 @@ - (void)peerConnection:(RTCPeerConnection *)peerConnection didChangeSignalingSta
...
+    [peerConnection addVideoTrackAdapter:streamReactTag track:track];
     [tracks addObject:@{@"id": track.trackId, @"kind": track.kind, @"label": track.trackId, @"enabled": @(track.isEnabled), @"remote": @(YES), @"readyState": @"live"}];
   }
   for (RTCAudioTrack *track in stream.audioTracks) {
     peerConnection.remoteTracks[track.trackId] = track;
     [tracks addObject:@{@"id": track.trackId, @"kind": track.kind, @"label": track.trackId, @"enabled": @(track.isEnabled), @"remote": @(YES), @"readyState": @"live"}];
   }
 
-  NSString *streamReactTag;
-  // Make sure ID does not exist across local and remote streams (for any peerConnection)
```
Comment (paweldomas) [inline ios/RCTWebRTC/WebRTCModule+RTCPeerConnection.m:397]: this goes away because it's unlikely that the conflict will happen ?

File: ios/RCTWebRTC/WebRTCModule+RTCPeerConnection.m [hunk 2]
```diff
@@ -462,6 +478,11 @@ - (void)peerConnection:(RTCPeerConnection*)peerConnection didOpenDataChannel:(RT
                                                   body:body];
 }
 
+- (void)peerConnection:(nonnull RTCPeerConnection *)peerConnection didRemoveIceCandidates:(nonnull NSArray<RTCIceCandidate *> *)candidates {
```
Comment (paweldomas) [inline ios/RCTWebRTC/WebRTCModule+RTCPeerConnection.m:488]: is this related to this PR ?

File: ios/RCTWebRTC/WebRTCModule+VideoTrackAdapter.h [hunk 3]
```diff
@@ -0,0 +1,13 @@
+
+#import "WebRTCModule.h"
+#import <WebRTC/RTCPeerConnection.h>
+
+@interface RTCPeerConnection (VideoTrackAdapter)
+
+@property (nonatomic, strong) NSMutableDictionary<NSString *,  id> *videoTrackAdapters;
+
+- (void)addVideoTrackAdapter:(NSString*)streamReadtId track:(RTCVideoTrack*)track;
```
Comment (paweldomas) [inline ios/RCTWebRTC/WebRTCModule+VideoTrackAdapter.h]: what does "streamReadtId" stand for ? is there a typo ?

Comment (saghul) [reply ios/RCTWebRTC/WebRTCModule+VideoTrackAdapter.h]: It is supposed to be `streamReactId`, it's a typo. Will fix.

File: ios/RCTWebRTC/WebRTCModule+VideoTrackAdapter.m [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+    _disposed = YES;
+    if (_timer != nil) {
+        dispatch_source_cancel(_timer);
+        _timer = nil;
+    }
+}
+
+- (void)emitMuteEvent:(BOOL)muted {
+    [self.module.bridge.eventDispatcher
+        sendDeviceEventWithName:@"peerConnectionSignalingStateChanged"
```
Comment (paweldomas) [inline ios/RCTWebRTC/WebRTCModule+VideoTrackAdapter.m]: This event name is incorrect, so the whole thing does not work, but I'll fix it since I'm at it.

File: RTCPeerConnection.js [hunk 5]
```diff
@@ -298,6 +307,17 @@ export default class RTCPeerConnection extends EventTarget(PEER_CONNECTION_EVENT
         }
         this.dispatchEvent(new MediaStreamEvent('removestream', {stream}));
       }),
+      DeviceEventEmitter.addListener('mediaStreamTrackMuteChanged', ev => {
```
Comment (paweldomas) [inline RTCPeerConnection.js:310]: this is missing in the iOS implementation

**Review by paweldomas — CHANGES_REQUESTED** [review]
The iOS implementation does not work

File: ios/RCTWebRTC/WebRTCModule+VideoTrackAdapter.m (hunk 4 above)
Comment (paweldomas) [reply ios/RCTWebRTC/WebRTCModule+VideoTrackAdapter.m]: it appears that just changing the name will not fix the issue, see other comment below about missing part

File: ios/RCTWebRTC/WebRTCModule+RTCPeerConnection.m (hunk 1 above)
Comment (saghul) [reply ios/RCTWebRTC/WebRTCModule+RTCPeerConnection.m:397]: More impossible than unlikely, it's a UUIDv4.

File: ios/RCTWebRTC/WebRTCModule+RTCPeerConnection.m (hunk 2 above)
Comment (saghul) [reply ios/RCTWebRTC/WebRTCModule+RTCPeerConnection.m:488]: Not really, it fixes a warning though.

File: ios/RCTWebRTC/WebRTCModule+VideoTrackAdapter.m (hunk 4 above)
Comment (saghul) [reply ios/RCTWebRTC/WebRTCModule+VideoTrackAdapter.m]: what comments below? I can't see any about missing stuff.

**Comment by saghul** [comment]
@paweldomas Fixed the typos / broken event name on iOS, sorry about that.

I remember testing this was... challenging. So I simulated it by removing the counter increase in the adapter, thus exercising the logic as if no frames were received.

I don't see why it wouldn't work if the emitted event name is correct (I fixed that now). Can you please give me some more details on what you tried and observed?

**Comment by paweldomas** [comment]
On Android there is "DeviceEventEmitter.addListener('mediaStreamTrackMuteChanged'"
There is no handler for iOS for this event which needs to go as a track event (it's possible that it could be emitted directly without the extra handler).

**Comment by paweldomas** [comment]
I'm testing that by blocking port which goes to the JVB from one of the participants. You need to see RTC mute event log on the JavaScript side (which you will see on the Android side).

**Comment by paweldomas** [comment]
https://github.com/jitsi/react-native-webrtc/pull/17/files/ae4666d1425c58858d8d9be4d198174efc7f915b#diff-a8c582915b8d289e3e2ade78c6b7c915R310

**Comment by saghul** [comment]
> On Android there is "DeviceEventEmitter.addListener('mediaStreamTrackMuteChanged'"
> There is no handler for iOS for this event which needs to go as a track event (it's possible that it could be emitted directly without the extra handler).

I don't understand what you mean here. That code is in JS, which runs on both. The event name was wrong, yes, that is fixed now. So what's missing in iOS then? I really can't see it.

**Comment by paweldomas** [comment]
> I don't understand what you mean here. That code is in JS, which runs on both. The event name was wrong, yes, that is fixed now. So what's missing in iOS then? I really can't see it.

🤦‍♂️  yes, you're right. Even with the new name it was not working for me, but I'll give it another try

File: ios/RCTWebRTC/WebRTCModule+RTCPeerConnection.m (hunk 1 above)
Comment (paweldomas) [reply ios/RCTWebRTC/WebRTCModule+RTCPeerConnection.m:397]: even if those are generated on different machines (remote streams) ?

Comment (saghul) [reply ios/RCTWebRTC/WebRTCModule+RTCPeerConnection.m:397]: Yes. https://en.wikipedia.org/wiki/Universally_unique_identifier#Version_4_(random) see the last sentence.

File: ios/RCTWebRTC/WebRTCModule+VideoTrackAdapter.h (hunk 3 above)
Comment (paweldomas) [reply ios/RCTWebRTC/WebRTCModule+VideoTrackAdapter.h]: is this typo still there or something's wrong with Github or am I blind ?

**Comment by paweldomas** [comment]
@saghul I guess you haven't pushed the fixes yet, right ?

**Comment by saghul** [comment]
@paweldomas I had pushed them to the wrong remote 🤦‍♂️  Here you go.

**Comment by paweldomas** [comment]
It's still not working for me, there's no video mute event on the JS side...

**Comment by saghul** [comment]
Damn. Let’s talk tomorrow about how you’ve been testing and I’ll debug it further.

Thanks for your efforts so far!

**Comment by saghul** [comment]
@paweldomas You're going to have to forgive me man, I'm an idiot. The problem was this line was missing: https://github.com/jitsi/react-native-webrtc/pull/17/files#diff-37692466c9d1561b7081c682a81fd67fR98 thanks to the magic of Objective-C everything seemed to work despite module being null. So the event was never sent to the JS side indeed.

In order to test this easily I commented out this line: https://github.com/jitsi/react-native-webrtc/pull/17/files#diff-0268bba36ea56f334811ecba6dc243eaR123 so the frame count will forever remain in 0 and after a bit (~3 seconds I think) you'll see the "XXX is having connectivity issues" message.

I'm sorry for the trouble.

PS: This PR also needs to be applied or there will be no UI change: https://github.com/jitsi/lib-jitsi-meet/pull/674 (I just rebased it)

**Review by paweldomas — APPROVED** [review]
It's working on both Android and iOS. Good job !
